File: Midterm1/Q4.py
'''
Question4: (S1, S3 & S4)
A. Fit the data using gussian function 
'''
import ROOT


ROOT.gROOT.SetStyle("ATLAS")
ROOT.gROOT.ForceStyle()

ROOT.gStyle.SetLabelFont(22, "X")
ROOT.gStyle.SetLabelFont(22, "Y")
ROOT.gStyle.SetTitleFont(22, "X")
ROOT.gStyle.SetTitleFont(22, "Y")

ROOT.gStyle.SetLabelSize(0.03, "X")
ROOT.gStyle.SetLabelSize(0.03, "Y")
ROOT.gStyle.SetTitleSize(0.04, "X")
ROOT.gStyle.SetTitleSize(0.04, "Y")
ROOT.gStyle.SetMarkerSize(0.5)

ROOT.gStyle.SetLegendFont(22)

file_path = "mlpHiggs.root"
file = ROOT.TFile(file_path, "READ")

bg_tree = file.Get("bg_filtered")

bg_hist = ROOT.TH1F("", "", 150, 0, 0.5)

for j in range(bg_tree.GetEntries()):
    bg_tree.GetEntry(j)
    bg_hist.Fill(bg_tree.ptsumf)

bg_hist.SetStats(0)

canvas = ROOT.TCanvas("canvas", "", 800, 600)
bg_hist.GetXaxis().SetTitle("Gaussian Fit_background")
bg_hist.GetYaxis().SetTitle("Evens")
bg_hist.Draw("HIST")

gauss= ROOT.TF1("gauss","gaus",0,0.5)
gauss.SetParLimits(0,40,1)
bg_hist.Fit(gauss)
gauss.SetLineColor(ROOT.kRed)
gauss.Draw("SAME")
mean = gauss.GetParameter(1)
sigma = gauss.GetParameter(2)
mean_error = gauss.GetParError(1)
sigma_error = gauss.GetParError(2)
legend = ROOT.TLegend(0.7, 0.7, 0.9, 0.9)
legend.SetHeader("made by Aishah")
legend.AddEntry(bg_hist, "background", "l")
legend.AddEntry(gauss, "Gaussian Fit", "l")
legend.AddEntry("", f"#mu: {mean:.2f} #pm {mean_error:.3f}", "")
legend.AddEntry("", f"#sigma: {sigma:.2f} #pm {sigma_error:.3f}", "")
legend.Draw()
canvas.SaveAs("Q4.pdf")
canvas.Update()
ROOT.gApplication.Run()
file.Close()

File: Finel_Exame/Q4.cpp

//Question-4 S(CLO 2.3)-10points:
/*
Fitting a Histogram with a Gaussian Function and Displaying Parameters
1. Fit the Histogram:
Use the TH1::Fit method to fit the histogram data_hist with a Gaussian function.
2. Retrieve Fit Parameters:
After fitting, obtain the Gaussian parameters: amplitude, mean, and sigma.
3. Update the Legend:
Modify the legend to include the fit parameters with their values and uncertainties.
*/


void Q4(){
  
  gStyle->SetLabelFont(22, "X");
  gStyle->SetLabelFont(22, "Y");
    gStyle->SetTitleFont(22, "X");
    gStyle->SetTitleFont(22, "Y");

    gStyle->SetLabelSize(0.2, "X");
    gStyle->SetLabelSize(0.03, "Y");
    gStyle->SetTitleSize(0.02, "X");
    gStyle->SetTitleSize(0.04, "Y");
    gStyle->SetMarkerSize(0.5);
  
    
  TFile* file = TFile::Open("histo.root","READ");
  TH1* data_hist = (TH1*)file->Get("data");
  TH1* mc_hist = (TH1*)file->Get("MC");

  TCanvas* canvas = new TCanvas("","", 800, 600);
  canvas->SetMargin(0.15,0.05,0.15,0.1);
  data_hist->GetXaxis()->SetTitle(" mass");
  data_hist->GetYaxis()->SetTitle("gauss_data");
  data_hist->SetMarkerColor(kBlue);
  mc_hist->SetMarkerColor(kRed);
  data_hist->SetStats(0);
  mc_hist->SetStats(0);
  data_hist->Draw("P");
  TF1* gauss= new TF1("gauss","gaus",-5,5);
  gauss->SetParLimits(0,230,300);
  data_hist->Fit("gauss");
  gauss->SetLineColor(kRed);
 
  
  
 
   double amplitude = gauss->GetParameter(0);
  double mean = gauss->GetParameter(1);
  double sigma = gauss->GetParameter(2);
  double mean_error = gauss->GetParError(1);
  double sigma_error= gauss->GetParError(2);

  TLegend* legend = new TLegend(0.5, 0.5, 0.7, 0.7);
  legend->SetHeader("made by AISHAH");
  legend->AddEntry(data_hist, "Data", "P");
  legend->AddEntry((TObject*)0,Form("#mu: %.2f #pm %.3f",mean, mean_error ),"");
    legend->AddEntry((TObject*)0,Form("#sigma: %.2f #pm %.3f",sigma, sigma_error ),"");
  legend->Draw();

  //  data_hist->GetYaxis()->SetRangeUser(0.0,220);
  canvas->SaveAs("Plots/Data_MC_Histograms.png");
  canvas->Show();
  //  canvas->Update();
}




    
